src/orange/surf/SoftSurfaceEqual.cc

```diff
@@ -0,0 +1,173 @@
+//---------------------------------------------------------------------------//
+//! \file orange/surf/SoftSurfaceEqual.cc
+//---------------------------------------------------------------------------//
+#include "SoftSurfaceEqual.hh"
+
+#include <cmath>
+
+#include "detail/AllSurfaces.hh"
+
+namespace celeritas
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+#define ORANGE_INSTANTIATE_OP(S)                                         \
+    template bool SoftSurfaceEqual::operator()(S<Axis::x> const&,        \
+                                               S<Axis::x> const&) const; \
+    template bool SoftSurfaceEqual::operator()(S<Axis::y> const&,        \
+                                               S<Axis::y> const&) const; \
+    template bool SoftSurfaceEqual::operator()(S<Axis::z> const&,        \
+                                               S<Axis::z> const&) const
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two aligned planes for near equality.
+ */
+template<Axis T>
+bool SoftSurfaceEqual::operator()(PlaneAligned<T> const& a,
+                                  PlaneAligned<T> const& b) const
+{
+    return soft_eq_(a.position(), b.position());
+}
+
+//! \cond
+ORANGE_INSTANTIATE_OP(PlaneAligned);
+//! \endcond
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two centered axis-aligned cylinders for near equality.
+ */
+template<Axis T>
+bool SoftSurfaceEqual::operator()(CylCentered<T> const& a,
+                                  CylCentered<T> const& b) const
+{
+    return soft_eq_sq_(a.radius_sq(), b.radius_sq());
+}
+
+//! \cond
+ORANGE_INSTANTIATE_OP(CylCentered);
+//! \endcond
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two centered spheres for near equality.
+ */
+bool SoftSurfaceEqual::operator()(SphereCentered const& a,
+                                  SphereCentered const& b) const
+{
+    return soft_eq_sq_(a.radius_sq(), b.radius_sq());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two aligned cylinders for near equality.
+ */
+template<Axis T>
+bool SoftSurfaceEqual::operator()(CylAligned<T> const& a,
+                                  CylAligned<T> const& b) const
+{
+    return soft_eq_sq_(a.radius_sq(), b.radius_sq())
+           && soft_eq_distance_(a.calc_origin(), b.calc_origin());
+}
+
+//! \cond
+ORANGE_INSTANTIATE_OP(CylAligned);
+//! \endcond
+
+//---------------------------------------------------------------------------//
+/*!
+ * Compare two planes for near equality.
+ */
+bool SoftSurfaceEqual::operator()(Plane const& a, Plane const& b) const
+{
+    // Guard against dot product being slightly greater than 1 due to fp
```

Comment:
Might it be slightly less than 1 as well?

Reply:
Less than 1 wouldn't cause `sqrt` to be out of range, but this has been removed.